File: apps/cli/internal/api/methods.go
package api

type CreateOrganizationInput struct {
	Name          string
	MemberUserIDs []string
}

type CreateProjectInput struct {
	Name           string
	SourceTypeHint string
	RepoURL        string
	NodeID         string
	LocalPath      string
}

type RegisterNodeInput struct {
	NodeID         string
	Name           string
	Endpoint       string
	Metadata       map[string]any
	Scope          string
	UpdateIfExists *bool
}

type CreateWorkspaceInput struct {
	ID           string
	NodeID       string
	LocalPath    string
	Kind         string
	Branch       string
	SourceBranch string
}

type CloseWorkspaceInput struct {
	NodeID    string
	LocalPath string
	Kind      string
	Branch    string
}

func (c *Client) Health() (HealthResponse, error) {
	var response HealthResponse
	err := c.DoDecode("GET", "/health", nil, &response)
	return response, err
}

func (c *Client) WhoAmI() (MeResponse, error) {
	var response MeResponse
	err := c.DoDecode("GET", "/me", nil, &response)
	return response, err
}

func (c *Client) ListOrganizations() (ListOrganizationsResponse, error) {
	var response ListOrganizationsResponse
	err := c.DoDecode("GET", "/orgs", nil, &response)
	return response, err
}

func (c *Client) CreateOrganization(input CreateOrganizationInput) (CreateOrganizationResponse, error) {
	var response CreateOrganizationResponse
	err := c.DoDecode("POST", "/orgs", map[string]any{
		"name":          input.Name,
		"memberUserIds": input.MemberUserIDs,
	}, &response)
	return response, err
}

func (c *Client) DeleteOrganization(orgID string) (OKResponse, error) {
	var response OKResponse
	err := c.DoDecode("DELETE", "/orgs/"+orgID, nil, &response)
	return response, err
}

func (c *Client) AddOrganizationMember(orgID string, userID string, role string) (AddOrganizationMemberResponse, error) {
	var response AddOrganizationMemberResponse
	err := c.DoDecode("POST", "/orgs/"+orgID+"/members", map[string]string{
		"userId": userID,
		"role":   role,
	}, &response)
	return response, err
}

func (c *Client) RemoveOrganizationMember(orgID string, userID string) (OKResponse, error) {
	var response OKResponse
	err := c.DoDecode("DELETE", "/orgs/"+orgID+"/members/"+userID, nil, &response)
	return response, err
}

func (c *Client) ListNodes(orgID string) (ListNodesResponse, error) {
	var response ListNodesResponse
	err := c.DoDecode("GET", "/orgs/"+orgID+"/nodes", nil, &response)
	return response, err
}

func (c *Client) DeleteNode(orgID string, nodeID string) (OKResponse, error) {
	var response OKResponse
	err := c.DoDecode("DELETE", "/orgs/"+orgID+"/nodes/"+nodeID, nil, &response)
	return response, err
}

func (c *Client) RegisterNode(input RegisterNodeInput) (RegisterNodeResponse, error) {
	payload := map[string]any{
		"nodeId": input.NodeID,
		"name":   input.Name,
		"scope":  input.Scope,
	}
	if input.Endpoint != "" {
		payload["endpoint"] = input.Endpoint
	}
	if len(input.Metadata) > 0 {
		payload["metadata"] = input.Metadata
	}
	if input.UpdateIfExists != nil {
		payload["updateIfExists"] = *input.UpdateIfExists
	}

	var response RegisterNodeResponse
	err := c.DoDecode("POST", "/nodes/register", payload, &response)
	return response, err
}

func (c *Client) DeleteProject(orgID string, projectID string) (OKResponse, error) {
	var response OKResponse
	err := c.DoDecode("DELETE", "/orgs/"+orgID+"/projects/"+projectID, nil, &response)
	return response, err
}

func (c *Client) ListProjects(orgID string) (ListProjectsResponse, error) {
	var response ListProjectsResponse
	err := c.DoDecode("GET", "/orgs/"+orgID+"/projects", nil, &response)
	return response, err
}

func (c *Client) CreateProject(orgID string, input CreateProjectInput) (CreateProjectResponse, error) {
	payload := map[string]string{
		"name": input.Name,
	}
	if input.SourceTypeHint != "" {
		payload["sourceTypeHint"] = input.SourceTypeHint
	}
	if input.RepoURL != "" {
		payload["repoUrl"] = input.RepoURL
	}
	if input.NodeID != "" {
		payload["nodeId"] = input.NodeID
	}
	if input.LocalPath != "" {
		payload["localPath"] = input.LocalPath
	}

	var response CreateProjectResponse
	err := c.DoDecode("POST", "/orgs/"+orgID+"/projects", payload, &response)
	return response, err
}

func (c *Client) ListWorkspaces(orgID string, projectID string) (ListWorkspacesResponse, error) {
	var response ListWorkspacesResponse
	err := c.DoDecode("GET", "/orgs/"+orgID+"/projects/"+projectID+"/workspaces", nil, &response)
	return response, err
}

func (c *Client) CreateWorkspace(orgID string, projectID string, input CreateWorkspaceInput) (CreateWorkspaceResponse, error) {
	payload := map[string]string{
		"nodeId":    input.NodeID,
		"localPath": input.LocalPath,
		"kind":      input.Kind,
	}
	if input.ID != "" {
		payload["id"] = input.ID
	}
	if input.Branch != "" {
		payload["branch"] = input.Branch
	}
	if input.SourceBranch != "" {
		payload["sourceBranch"] = input.SourceBranch
	}

	var response CreateWorkspaceResponse
	err := c.DoDecode("POST", "/orgs/"+orgID+"/projects/"+projectID+"/workspaces", payload, &response)
	return response, err
}

type UpsertWorkspacePullRequestInput struct {
	PrID        string         `json:"prId"`
	Title       string         `json:"title,omitempty"`
	URL         string         `json:"url,omitempty"`
	Branch      string         `json:"branch,omitempty"`
	BaseBranch  string         `json:"baseBranch,omitempty"`
	State       string         `json:"state"`
	Metadata    map[string]any `json:"metadata,omitempty"`
	DetectedAt  string         `json:"detectedAt"`
	ResolvedAt  string         `json:"resolvedAt,omitempty"`
}

func (c *Client) UpsertWorkspacePullRequest(orgID string, projectID string, workspaceID string, input UpsertWorkspacePullRequestInput) (OKResponse, error) {
	var response OKResponse
	err := c.DoDecode("PUT", "/orgs/"+orgID+"/projects/"+projectID+"/workspaces/"+workspaceID+"/pull-requests", input, &response)
	return response, err
}

func (c *Client) CloseWorkspace(orgID string, projectID string, input CloseWorkspaceInput) (CreateWorkspaceResponse, error) {
	payload := map[string]string{
		"nodeId":    input.NodeID,
		"localPath": input.LocalPath,
		"kind":      input.Kind,
	}
	if input.Branch != "" {
		payload["branch"] = input.Branch
	}

	var response CreateWorkspaceResponse
	err := c.DoDecode("PATCH", "/orgs/"+orgID+"/projects/"+projectID+"/workspaces/close", payload, &response)
	return response, err
}

func (c *Client) RelayToken(nodeID string) (RelayTokenResponse, error) {
	var response RelayTokenResponse
	err := c.DoDecode("POST", "/nodes/"+nodeID+"/relay-token", nil, &response)
	return response, err
}

func (c *Client) RefreshToken(refreshToken string) (RefreshTokenResponse, error) {
	var response RefreshTokenResponse
	err := c.DoDecode("POST", "/auth/refresh", map[string]string{
		"refreshToken": refreshToken,
	}, &response)
	return response, err
}

func (c *Client) RevokeToken(refreshToken string) (OKResponse, error) {
	var response OKResponse
	err := c.DoDecode("POST", "/auth/revoke", map[string]string{
		"refreshToken": refreshToken,
	}, &response)
	return response, err
}

type StartScheduledJobRunInput struct {
	RunID     string
	StartedAt string
}

type CompleteScheduledJobRunInput struct {
	RunID        string
	FinishedAt   string
	Status       string
	ResponseBody string
	ErrorCode    string
	ErrorMessage string
	ErrorDetails map[string]any
}

func (c *Client) StartScheduledJobRun(nodeID string, input StartScheduledJobRunInput) (OKResponse, error) {
	payload := map[string]any{
		"runId": input.RunID,
	}
	if input.StartedAt != "" {
		payload["startedAt"] = input.StartedAt
	}

	var response OKResponse
	err := c.DoDecode("PUT", "/nodes/"+nodeID+"/scheduled-jobs/runs/start", payload, &response)
	return response, err
}

func (c *Client) CompleteScheduledJobRun(nodeID string, input CompleteScheduledJobRunInput) (OKResponse, error) {
	payload := map[string]any{
		"runId":  input.RunID,
		"status": input.Status,
	}
	if input.FinishedAt != "" {
		payload["finishedAt"] = input.FinishedAt
	}
	if input.ResponseBody != "" {
		payload["responseBody"] = input.ResponseBody
	}
	if input.ErrorCode != "" {
		payload["errorCode"] = input.ErrorCode
	}
	if input.ErrorMessage != "" {
		payload["errorMessage"] = input.ErrorMessage
	}
	if len(input.ErrorDetails) > 0 {
		payload["errorDetails"] = input.ErrorDetails
	}

	var response OKResponse
	err := c.DoDecode("PUT", "/nodes/"+nodeID+"/scheduled-jobs/runs/complete", payload, &response)
	return response, err
}

File: apps/cli-rs/src/api/methods.rs
use super::client::ApiClient;
use super::types::*;
use reqwest::Method;
use serde_json::json;
use std::collections::HashMap;

impl ApiClient {
    // ── System ────────────────────────────────────────────────────────────────

    pub async fn health(&self) -> anyhow::Result<HealthResponse> {
        self.do_decode::<_, HealthResponse>(Method::GET, "/health", None::<&()>).await
    }

    pub async fn whoami(&self) -> anyhow::Result<MeResponse> {
        self.do_decode::<_, MeResponse>(Method::GET, "/me", None::<&()>).await
    }

    // ── Auth ──────────────────────────────────────────────────────────────────

    pub async fn refresh_token(&self, refresh_token: &str) -> anyhow::Result<RefreshTokenResponse> {
        let body = json!({ "refreshToken": refresh_token });
        self.do_decode(Method::POST, "/auth/refresh", Some(&body)).await
    }

    pub async fn revoke_token(&self, refresh_token: &str) -> anyhow::Result<OkResponse> {
        let body = json!({ "refreshToken": refresh_token });
        self.do_decode(Method::POST, "/auth/revoke", Some(&body)).await
    }

    // ── Service Tokens ────────────────────────────────────────────────────────

    pub async fn list_service_tokens(&self) -> anyhow::Result<ListServiceTokensResponse> {
        self.do_decode::<_, _>(Method::GET, "/service-tokens", None::<&()>).await
    }

    pub async fn create_service_token(
        &self,
        name: &str,
        expires_in_days: Option<u32>,
    ) -> anyhow::Result<CreateServiceTokenResponse> {
        let mut body = json!({ "name": name });
        if let Some(days) = expires_in_days {
            body["expiresInDays"] = json!(days);
        }
        self.do_decode(Method::POST, "/service-tokens", Some(&body)).await
    }

    pub async fn revoke_service_token(&self, token_id: &str) -> anyhow::Result<OkResponse> {
        let path = format!("/service-tokens/{token_id}");
        self.do_decode::<_, OkResponse>(Method::DELETE, &path, None::<&()>).await
    }

    // ── Organizations ─────────────────────────────────────────────────────────

    pub async fn list_organizations(&self) -> anyhow::Result<ListOrganizationsResponse> {
        self.do_decode::<_, _>(Method::GET, "/orgs", None::<&()>).await
    }

    pub async fn create_organization(
        &self,
        name: &str,
        member_user_ids: Vec<String>,
    ) -> anyhow::Result<CreateOrganizationResponse> {
        let body = json!({ "name": name, "memberUserIds": member_user_ids });
        self.do_decode(Method::POST, "/orgs", Some(&body)).await
    }

    pub async fn delete_organization(&self, org_id: &str) -> anyhow::Result<OkResponse> {
        let path = format!("/orgs/{org_id}");
        self.do_decode::<_, OkResponse>(Method::DELETE, &path, None::<&()>).await
    }

    pub async fn add_organization_member(
        &self,
        org_id: &str,
        user_id: &str,
        role: &str,
    ) -> anyhow::Result<AddOrganizationMemberResponse> {
        let path = format!("/orgs/{org_id}/members");
        let body = json!({ "userId": user_id, "role": role });
        self.do_decode(Method::POST, &path, Some(&body)).await
    }

    pub async fn remove_organization_member(
        &self,
        org_id: &str,
        user_id: &str,
    ) -> anyhow::Result<OkResponse> {
        let path = format!("/orgs/{org_id}/members/{user_id}");
        self.do_decode::<_, OkResponse>(Method::DELETE, &path, None::<&()>).await
    }

    // ── Nodes ─────────────────────────────────────────────────────────────────

    pub async fn list_nodes(&self, org_id: &str) -> anyhow::Result<ListNodesResponse> {
        let path = format!("/orgs/{org_id}/nodes");
        self.do_decode::<_, _>(Method::GET, &path, None::<&()>).await
    }

    pub async fn delete_node(&self, org_id: &str, node_id: &str) -> anyhow::Result<OkResponse> {
        let path = format!("/orgs/{org_id}/nodes/{node_id}");
        self.do_decode::<_, OkResponse>(Method::DELETE, &path, None::<&()>).await
    }

    pub async fn update_node_scope(
        &self,
        org_id: &str,
        node_id: &str,
        scope: &str,
    ) -> anyhow::Result<UpdateNodeScopeResponse> {
        let path = format!("/orgs/{org_id}/nodes/{node_id}/scope");
        let body = json!({ "scope": scope });
        self.do_decode(Method::PATCH, &path, Some(&body)).await
    }

    pub async fn register_node(
        &self,
        node_id: &str,
        name: &str,
        kind: Option<&str>,
        endpoint: Option<&str>,
        metadata: Option<HashMap<String, serde_json::Value>>,
        scope: &str,
        update_if_exists: Option<bool>,
    ) -> anyhow::Result<RegisterNodeResponse> {
        let mut body = json!({
            "nodeId": node_id,
            "name": name,
            "scope": scope,
        });
        if let Some(k) = kind.filter(|s| !s.is_empty()) {
            body["kind"] = json!(k);
        }
        if let Some(e) = endpoint.filter(|s| !s.is_empty()) {
            body["endpoint"] = json!(e);
        }
        if let Some(m) = metadata.filter(|m| !m.is_empty()) {
            body["metadata"] = json!(m);
        }
        if let Some(u) = update_if_exists {
            body["updateIfExists"] = json!(u);
        }
        self.do_decode(Method::POST, "/nodes/register", Some(&body)).await
    }

    // ── Projects ──────────────────────────────────────────────────────────────

    pub async fn list_projects(&self, org_id: &str) -> anyhow::Result<ListProjectsResponse> {
        let path = format!("/orgs/{org_id}/projects");
        self.do_decode::<_, _>(Method::GET, &path, None::<&()>).await
    }

    pub async fn create_project(
        &self,
        org_id: &str,
        name: &str,
        source_type_hint: Option<&str>,
        repo_url: Option<&str>,
        node_id: Option<&str>,
        local_path: Option<&str>,
    ) -> anyhow::Result<CreateProjectResponse> {
        let mut body = json!({ "name": name });
        if let Some(v) = source_type_hint.filter(|s| !s.is_empty()) {
            body["sourceTypeHint"] = json!(v);
        }
        if let Some(v) = repo_url.filter(|s| !s.is_empty()) {
            body["repoUrl"] = json!(v);
        }
        if let Some(v) = node_id.filter(|s| !s.is_empty()) {
            body["nodeId"] = json!(v);
        }
        if let Some(v) = local_path.filter(|s| !s.is_empty()) {
            body["localPath"] = json!(v);
        }
        let path = format!("/orgs/{org_id}/projects");
        self.do_decode(Method::POST, &path, Some(&body)).await
    }

    pub async fn delete_project(
        &self,
        org_id: &str,
        project_id: &str,
    ) -> anyhow::Result<OkResponse> {
        let path = format!("/orgs/{org_id}/projects/{project_id}");
        self.do_decode::<_, OkResponse>(Method::DELETE, &path, None::<&()>).await
    }

    // ── Workspaces ────────────────────────────────────────────────────────────

    pub async fn list_workspaces(
        &self,
        org_id: &str,
        project_id: &str,
    ) -> anyhow::Result<ListWorkspacesResponse> {
        let path = format!("/orgs/{org_id}/projects/{project_id}/workspaces");
        self.do_decode::<_, _>(Method::GET, &path, None::<&()>).await
    }

    pub async fn create_workspace(
        &self,
        org_id: &str,
        project_id: &str,
        id: Option<&str>,
        node_id: &str,
        local_path: &str,
        kind: &str,
        branch: Option<&str>,
        source_branch: Option<&str>,
    ) -> anyhow::Result<CreateWorkspaceResponse> {
        let mut body = json!({
            "nodeId": node_id,
            "localPath": local_path,
            "kind": kind,
        });
        if let Some(v) = id.filter(|s| !s.is_empty()) {
            body["id"] = json!(v);
        }
        if let Some(v) = branch.filter(|s| !s.is_empty()) {
            body["branch"] = json!(v);
        }
        if let Some(v) = source_branch.filter(|s| !s.is_empty()) {
            body["sourceBranch"] = json!(v);
        }
        let path = format!("/orgs/{org_id}/projects/{project_id}/workspaces");
        self.do_decode(Method::POST, &path, Some(&body)).await
    }

    pub async fn close_workspace(
        &self,
        org_id: &str,
        project_id: &str,
        workspace_id: &str,
    ) -> anyhow::Result<CreateWorkspaceResponse> {
        let path = format!("/orgs/{org_id}/projects/{project_id}/workspaces/close");
        let body = json!({ "workspaceId": workspace_id });
        self.do_decode(Method::PATCH, &path, Some(&body)).await
    }

    // ── Relay / Node tokens ───────────────────────────────────────────────────

    pub async fn relay_token(&self, node_id: &str) -> anyhow::Result<RelayTokenResponse> {
        let path = format!("/nodes/{node_id}/relay-token");
        self.do_decode::<_, _>(Method::POST, &path, None::<&()>).await
    }

    // ── Scheduled Jobs ────────────────────────────────────────────────────────

    pub async fn start_scheduled_job_run(
        &self,
        node_id: &str,
        run_id: &str,
        started_at: Option<&str>,
    ) -> anyhow::Result<OkResponse> {
        let mut body = json!({ "runId": run_id });
        if let Some(t) = started_at.filter(|s| !s.is_empty()) {
            body["startedAt"] = json!(t);
        }
        let path = format!("/nodes/{node_id}/scheduled-jobs/runs/start");
        self.do_decode(Method::PUT, &path, Some(&body)).await
    }

    pub async fn complete_scheduled_job_run(
        &self,
        node_id: &str,
        run_id: &str,
        status: &str,
        finished_at: Option<&str>,
        response_body: Option<&str>,
        error_code: Option<&str>,
        error_message: Option<&str>,
        error_details: Option<HashMap<String, serde_json::Value>>,
    ) -> anyhow::Result<OkResponse> {
        let mut body = json!({ "runId": run_id, "status": status });
        if let Some(t) = finished_at.filter(|s| !s.is_empty()) {
            body["finishedAt"] = json!(t);
        }
        if let Some(r) = response_body.filter(|s| !s.is_empty()) {
            body["responseBody"] = json!(r);
        }
        if let Some(c) = error_code.filter(|s| !s.is_empty()) {
            body["errorCode"] = json!(c);
        }
        if let Some(m) = error_message.filter(|s| !s.is_empty()) {
            body["errorMessage"] = json!(m);
        }
        if let Some(d) = error_details.filter(|d| !d.is_empty()) {
            body["errorDetails"] = json!(d);
        }
        let path = format!("/nodes/{node_id}/scheduled-jobs/runs/complete");
        self.do_decode(Method::PUT, &path, Some(&body)).await
    }

    // ── Token Usage ───────────────────────────────────────────────────────────

    pub async fn upsert_token_usage_hourly(
        &self,
        org_id: &str,
        rows: &[TokenUsageHourlyRow],
    ) -> anyhow::Result<OkResponse> {
        let path = format!("/orgs/{org_id}/token-usage/hourly");
        let body = json!({ "rows": rows });
        self.do_decode(Method::POST, &path, Some(&body)).await
    }
}
